Adds support for custom audits and gatherers

paullewis opened the pull request:
This PR allows us to specify additional paths for audits and gatherers in the config, which get searched when we do the requires.

The properties in the config are thus:

```json
{
  "paths": {
    "audits": [
      "/path/to/my/audits"
    ],

    "gatherers": [
      "/path/to/my/gatherers"
    ]
  }
}
```

A custom gatherer looks like this:

```javascript
/**
 *
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const LighthouseGatherer = require('lighthouse/custom/gatherer');

class MediaQueries extends LighthouseGatherer {
  pass(options) {
    const driver = options.driver;
    return driver.sendCommand('CSS.getMediaQueries')
        .then(result => {
          this.artifact = result
        })
  }
}

module.exports = MediaQueries;
```

and a custom audit like this:

```javascript
/**
 *
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const LighthouseAudit = require('lighthouse/custom/audit');

class MediaQueries extends LighthouseAudit {
  static get meta () {
    return {
      category: 'Custom',
      name: 'media-queries',
      description: 'Has media queries',
      requiredArtifacts: ['MediaQueries']
    };
  }

  static audit(artifacts) {
    return MediaQueries.generateAuditResult({
      rawValue: artifacts.MediaQueries.length > 0
    });
  }
}

module.exports = MediaQueries;
```

Then if you fancy it you can just adjust the reporting to suit.

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -171,16 +172,79 @@ function filterAudits(audits, auditWhitelist) {
+        return definition;
+      }
+
+      const requirePath = auditPath.startsWith('/') ? auditPath : path.join(rootPath, auditPath);
```
I think there's a path.resolve() or something to sort out absolute vs not. Would rather that then us handling the branch.

paulirish replied:
Yup: The last 3 here are exactly what we need:
![image](https://cloud.githubusercontent.com/assets/39191/17722617/57f16e78-63e8-11e6-8046-dd84dbd2c983.png)

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -171,16 +172,79 @@ function filterAudits(audits, auditWhitelist) {
-      return require(`../audits/${audit}`);
-    } catch (requireError) {
+    // Check each path to see if the audit can be located. First match wins.
+    const AuditClass = paths.reduce((definition, auditPath) => {
```
Let's give this section a name like `findAuditDefinitions` or something? At this point `expand` feels like a poor match.

paullewis replied:
Agreed.

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -171,16 +172,79 @@ function filterAudits(audits, auditWhitelist) {
+            return errorList + (auditValidation[item] ? '' : `\n - ${item} is missing`);
+          }, '');
+
+      throw new Error(`Invalid audit class: ${errors}`);
```
Can we clarify the audit was found but is deemed invalid?

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -171,16 +172,79 @@ function filterAudits(audits, auditWhitelist) {
 
-function expandAudits(audits) {
+function expandAudits(audits, paths) {
+  const rootPath = path.join(__dirname, '../../');
```
This would be more clear to me if there is one loop which resolves our core audits, and then a second one afterwards which hunts down ones not resolved in the provided paths.

paulirish replied:
That way if our auditList is fully hydrated, then we skip looking in these other paths.

brendankenny replied:
yeah, could just check in `Runner.getAuditList()`, if not found there, then go through the path search

paullewis replied:
So to be clear, we attempt a single pass in the "standard" / core location, and then a second pass looking in other locations?

paulirish commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -214,26 +221,55 @@ class GatherRunner {
+        return definition;
+      }
+
+      const requirePath = gathererPath.startsWith('/') ?
```
repeat: path.resolve() to do this for us

paullewis replied:
Yep

brendankenny commented on custom/audit.js:
```diff
@@ -0,0 +1,19 @@
+ */
+
+// Pass-through file for developer convenience.
+module.exports = require('../lighthouse-core/audits/audit');
```
is this so you can `class MyAudit extends require('custom/audit')`? I'd rather do just expose on the index.js module so you could do `class MyAudit extends require('lighthouse').Audit`

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -171,16 +172,79 @@ function filterAudits(audits, auditWhitelist) {
 }
 
+function validateAudit(auditDefinition) {
+  const hasAuditMethod = typeof auditDefinition.audit === 'function';
```
This method is pretty verbose. Wdyt about....

```js
function validateAudit(auditDefinition) {
  let results = {
    items: {
      'audit()': typeof auditDefinition.audit === 'function',
      'meta property': typeof auditDefinition.meta === 'object',
      'meta.name property': hasMeta && typeof auditDefinition.meta.name !== 'undefined',
      'meta.category property': hasMeta && typeof auditDefinition.meta.category !== 'undefined',
      'meta.description property': hasMeta && typeof auditDefinition.meta.description !== 'undefined',
      'meta.requiredArtifacts array': hasMeta && Array.isArray(auditDefinition.meta.requiredArtifacts),
      'generateAuditResult()': typeof auditDefinition.generateAuditResult === 'function'
    }
  };

  // true if all values of results.items are true
  results.valid = Object.values(results.items).reduce((cur, soFar) => soFar && cur, true);
  return results;
}
```

brendankenny replied:
one other idea: this method could also be more something like `assertValidAudit()` and throw directly instead of passing an object back out. Would skip a lot of the object building only to concat it all back together

paulirish replied:
Fair. Might as well throw as we don't really want to recover from this.

paullewis replied:
Agreed.

paulirish commented on lighthouse-core/config/index.js:
```diff
@@ -241,19 +305,53 @@ class Config {
-        filterAudits(configJSON.audits, auditWhitelist)
+    this._configJSON = this._initRequirePaths(configJSON);
+
+    this._audits = this.json.audits ? expandAudits(
```
Why do we continue if `this.json.audits` is falsy?   I understand for passes & artifacts, but doesn't really make sense here.

paullewis replied:
Because we already did. The only thing I changed here was to make it a getter on this.json. So if there's a bug / invalid approach, it already existed :)

brendankenny commented on lighthouse-core/config/index.js:
```diff
@@ -241,19 +305,53 @@ class Config {
 
-    this._audits = configJSON.audits ? expandAudits(
-        filterAudits(configJSON.audits, auditWhitelist)
+    this._configJSON = this._initRequirePaths(configJSON);
```
need to not modify `configJSON` in place as `_initRequirePaths` does, as the config is a live object and e.g. used repeatedly by the extension (I wonder if we should just have `Config` expect a string `configJSON` the it `JSON.parse`s so that it's always a new object and we can avoid being careful about this)

brendankenny replied:
why not just store it on `this.json` if you're just going to make a getter for `this.json`? :P

brendankenny commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -169,6 +170,12 @@ class GatherRunner {
       options.flags = {};
     }
 
+    if (typeof options.config === 'undefined') {
```
this check should be in `Runner`

paullewis replied:
Why? The Gatherer needs the config, it should be the one to assert that it exists.

brendankenny commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -214,26 +221,55 @@ class GatherRunner {
         const artifacts = Object.assign({}, tracingData);
         passes.forEach(pass => {
           pass.gatherers.forEach(gatherer => {
+            if (typeof gatherer.artifact === 'undefined') {
```
maybe `if ('artifact' in gatherer)` to account for gatherers that may someday want to return an undefined

paullewis replied:
I don't think I want a gatherer to return undefined as it's easily mistakable for a failure to set `artifact`. If you want to say error then a -1 would do it.

brendankenny commented:
This will be great to have (and love the audit validation :). Two more thoughts:

- Some discussion about this earlier: the custom audits and gatherers may need to be specified relative to the config file, as this is really the only fixed point we know (who knows how the user has installed and is running Lighthouse and so where config/index.js is relative to where they think the root of their project is).

- I wonder if this should use inline paths instead of an explicit `paths` property on the config. If there are many gatherers or audits in the same directory, it will cause more noise, but if there are many different directories (like a directory per audit e.g. from npm modules), the `paths` thing gets a bit unwieldy. For maintenance purposes it also may just be better to have path and module name all in one place.

The config would then look something like:
```
{
  "audits": [
    "is-on-https",
    "redirects-http",
    "service-worker",
    "path/to/my/audits/audit1/cool-audit",
    "audit-security-mixedcontent/audits/better-audit"
  ]
}
```
(similar for gatherers)

This would also support a more straightforward include process, basically
- get `Runner.getAuditList()`
- if the audit is in there, include it from `./audits/{$auditName}`
- otherwise adjust the path to be [relative](https://nodejs.org/api/path.html#path_path_relative_from_to) to `config/index.js`
- `require()` the new string

This lets `require` get us all the extra functionality like starting with `./` or `../` for relative paths, traversing folders to the root looking for `node_modules`, supporting `$NODE_PATH`, etc. The flexible `node_modules/` support I think will be particularly helpful and avoid having to write support in for many Lighthouse use corner cases ourselves.

It also sometimes eliminates the need for paths in the config file, since if they're in a `node_modules/` directory you can do the normal `require()` thing and drop the path, so it goes back to
```
{
  "audits": [
    "is-on-https",
    "redirects-http",
    "service-worker",
    "cool-audit",
    "better-audit"
  ]
}
```

paullewis commented:
Okay, hopefully I've struck the right balance here:

* Changed it so that we assume requires in this order: 1) core (checked through `Runner.getAuditList()`, 2) require directly (via `require.resolve`), 3) require relatively via the config path.
* Added checks for invalid gatherers as well as invalid audits.
* Removed the custom folder from the root; exported the Audit and Gatherer classes via the Lighthouse module, i.e. `require('lighthouse').Audit`

PTAL

paulirish commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -189,7 +190,29 @@ class GatherRunner {
+      return Promise.reject(new Error('You must provide a config'));
+    }
+
+    // Default mobile emulation and page loading to true.
```
This set of checks duplicates what's already in runner. : https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/runner.js#L29-L44

I think we should only check where neccessary and not end up with all the duplication. If we can't keep straight if we need a safety check, then we probably need to untangle a bit. (I'm thinking of our refactor for `traceEvents` a bit ago. There's a few more in the wings.)

paullewis replied:
I agree. This looks like an artifact of my refactoring perhaps, rather than something that _should_ be there.

paulirish commented:
lgtm

paulirish commented:
I do think we should get @ebidel @hsablonniere or @blaine to bang around with this API to make sure this works well.

paullewis commented:
Agreed. I'm very open to this changing with developer feedback. My feeling is we should merge, and then revisit it once folks can have a play with it and let us know how well it works for them.

hsablonniere commented:
I won't be able to play with it over the weekend. Maybe next week. This definitely seems interesting.

paulirish commented:
K merging so it's available. 

@paullewis let's get some notes in the readme on how to use it.

ebidel commented:
Been waiting for this to merge so I can give feedback.   Too early to tell
right now :)

On Fri, Aug 19, 2016, 6:55 PM Paul Irish <notifications@github.com> wrote:

> Merged #593 <https://github.com/GoogleChrome/lighthouse/pull/593>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleChrome/lighthouse/pull/593#event-761951763>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOigHUGahUYEhVeAqBb0X6QnFe7coFhks5qhl6dgaJpZM4JlloK>
> .
>
